cumulus/pallets/executive/src/lib.rs

```diff
@@ -372,7 +431,8 @@ where
 			AllPalletsWithSystem,
 			COnRuntimeUpgrade,
 		>::apply_extrinsic(uxt);
-		// TODO: collect the storage root.
+		// TODO: when the extrinsic fails, the storage root does not change, thus skip it?
```

Comment:
I think it makes sense to push an item regardless like you did here such that number of items always matches number of applied extrinsics

Reply:
Though comment on this method says it is outside of block execution function, are you sure storage is available here?

Reply:
Do you mean if the storage root can be calculated correctly after the inner frame-executive `apply_extrinsic`? I have run this branch locally, works as expected with my eyes. Writing tests for cirrus is becoming pressing :P.

Reply:
> Do you mean if the storage root can be calculated correctly after the inner frame-executive apply_extrinsic?

Not sure what do you refer to here, I'm confused.

My concern is about "Apply extrinsic outside of the block execution function." above. Not sure what does "outside" mean exactly here. It might be comment originated from some implementation detail or might mean that this method is not guaranteed to be called in runtime context, in which case storage will not work as one might expect it to.

Reply:
> are you sure storage is available here?

I was confused about the storage you were referring to.

>Apply extrinsic outside of the block execution function.

I think it could mean outside of the `execute_block` function. These methods have to be called in runtime context.